Zend/zend_lazy_objects.c

```diff
@@ -0,0 +1,705 @@
+	 * supports the following pattern:
+	 * if (zend_lazy_object_must_init(obj)) {
+	 *     instance = zend_lazy_object_init(obj);
+	 * }
+	 */
+	if (zend_lazy_object_initialized(obj)) {
+		ZEND_ASSERT(zend_object_is_lazy_proxy(obj));
+		zend_lazy_object_info *info = zend_lazy_object_get_info(obj);
+		ZEND_ASSERT(info->flags & ZEND_LAZY_OBJECT_INITIALIZED);
+		return info->u.instance;
+	}
+
+	zend_class_entry *ce = obj->ce;
+
+	if (UNEXPECTED(!(ce->ce_flags & ZEND_ACC_CONSTANTS_UPDATED))) {
+		if (UNEXPECTED(zend_update_class_constants(ce) != SUCCESS)) {
+			ZEND_ASSERT(EG(exception));
+			return NULL;
+		}
+	}
+
+	if (zend_object_is_lazy_proxy(obj)) {
+		return zend_lazy_object_init_proxy(obj);
+	}
+
+	zend_fcall_info_cache *initializer = zend_lazy_object_get_initializer_fcc(obj);
+
+	/* Prevent reentrant initialization */
+	OBJ_EXTRA_FLAGS(obj) &= ~IS_OBJ_LAZY;
+
+	/* Snapshot dynamic properties */
+	HashTable *properties_snapshot = obj->properties;
+	if (properties_snapshot) {
+		GC_TRY_ADDREF(properties_snapshot);
+	}
+
+	zval *properties_table_snapshot = NULL;
+
+	/* Snapshot declared properties and initialize lazy properties to their
+	 * default value */
+	if (ce->default_properties_count) {
+		zval *default_properties_table = CE_DEFAULT_PROPERTIES_TABLE(ce);
+		zval *properties_table = obj->properties_table;
+		properties_table_snapshot = emalloc(sizeof(*properties_table_snapshot) * ce->default_properties_count);
+
+		for (int i = 0; i < ce->default_properties_count; i++) {
+			ZVAL_COPY_PROP(&properties_table_snapshot[i], &properties_table[i]);
+			if (Z_PROP_FLAG_P(&properties_table[i]) & IS_PROP_LAZY) {
+				ZVAL_COPY_PROP(&properties_table[i], &default_properties_table[i]);
+			}
+		}
+	}
+
+	/* Call initializer */
+	zval retval;
+	int argc = 1;
+	zval zobj;
+	HashTable *named_params = NULL;
+
+	ZVAL_OBJ(&zobj, obj);
+
+	zend_call_known_fcc(initializer, &retval, argc, &zobj, named_params);
+
+	if (EG(exception)) {
+		zend_lazy_object_revert_init(obj, properties_table_snapshot, properties_snapshot);
+		return NULL;
+	}
+
+	if (Z_TYPE(retval) != IS_NULL) {
+		zend_lazy_object_revert_init(obj, properties_table_snapshot, properties_snapshot);
+		zval_ptr_dtor(&retval);
+		zend_type_error("Lazy object initializer must return NULL or no value");
+		return NULL;
+	}
+
+	if (properties_table_snapshot) {
+		for (int i = 0; i < obj->ce->default_properties_count; i++) {
+			zval *p = &properties_table_snapshot[i];
+			/* Use zval_ptr_dtor directly here (not zend_object_dtor_property),
+			 * as any reference type_source will have already been deleted in
+			 * case the prop is not bound to this value anymore. */
+			i_zval_ptr_dtor(p);
+		}
+		efree(properties_table_snapshot);
+	}
+
+	if (properties_snapshot) {
+		zend_release_properties(properties_snapshot);
+	}
+
+	zend_lazy_object_del_info(obj);
+
+	return obj;
+}
+
+/* Mark an object as non-lazy (after all properties were initialized) */
+void zend_lazy_object_realize(zend_object *obj)
```

Comment:
It is necessary to use `ZEND_API` to expose functions to ext/reflection? My understanding of `ZEND_API` is that it is required to expose symbols outside of the current object, so usually we need it to expose functions to extensions. But reflection is always built in the main object, so it should be ok? Or should we use `ZEND_API` for consistency here?

Reply:
Indeed. But do we need it on zend_lazy_object_mark_as_initialized() then?

Reply:
We don't really need it, but I though it would be nice to expose a few high level functions that correspond to the userland API (`zend_lazy_object_mark_as_initialized()` implements `ReflectionClass::markLazyObjectAsInitialized()`). I've grouped them before other functions in the header file.

Reply:
Ok! I was just wondering if it was intentional or accidental, given that the two functions are pretty similar.